### PR TITLE
Update on mocking

### PR DESCRIPTION
**Pull request by maelle**

### CODE REVIEW

File: content/post/2024-03-12-mocking/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+## Mocking yay or nay
+
+The R packages book by Hadley Wickham and Jenny Bryan contains an insightful paragraph on [test coverage](https://r-pkgs.org/testing-design.html#sec-testing-design-coverage)[^1], including those lines:
+
+[^1]: The three chapters on testing are definitely worth a read!
+
+> "In many cases, that last 10% or 1% often requires some awkward gymnastics to cover. Sometimes this forces you to introduce mocking or some other new complexity. Don't sacrifice the maintainability of your test suite in the name of covering some weird edge case that hasn't yet proven to be a problem."
+
+This is definitely good to keep in mind whilst diving into the world of mocking.
+Not that mocking needs to be especially complex!
```
Comment (hfrick) [inline content/post/2024-03-12-mocking/index.Rmd:58]:
```suggestion
Not that mocking needs to be especially complex though!
```
I'm not entirely sure about this suggestion but somehow my brain struggled with this sentence when I first read it.

File: content/post/2024-03-12-mocking/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+Beside, sometimes, testing is complex because your code is begging to be refactored.
```
Comment (hfrick) [inline content/post/2024-03-12-mocking/index.Rmd:60]: I fee like you might have a link for this up your sleeve to let curious minds read on?

File: content/post/2024-03-12-mocking/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+:wink:
+
+## Revisiting the general mocking example
+
+The [example of general mocking](/2019/10/29/mocking/#general-mocking) in the post from 2019 can be rewritten to take advantage of the new mocking functionality that was [announced as no longer experimental in October 2023](https://www.tidyverse.org/blog/2023/10/testthat-3-2-0/#mocking).
```
Comment (hfrick) [inline content/post/2024-03-12-mocking/index.Rmd:65]: Is the passive voice here a conscious decision?

File: content/post/2024-03-12-mocking/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+  }
+
+  FALSE
+}
+```
+
+The function returns `TRUE` on Windows, `FALSE` on other operating systems.
+
+To test it no matter the operating system, we need to make our tests believe they are run on a given operating system.
+We can in this day and age achieve this using mocking tools from testthat itself!
```
Comment (hfrick) [inline content/post/2024-03-12-mocking/index.Rmd:101]:
```suggestion
We can now achieve this using mocking tools from testthat itself!
```
or put commas around "in this day and age"?

File: content/post/2024-03-12-mocking/index.Rmd [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+  expect_false(is_encoding_a_pain())
+})
+
+test_that("is_encoding_a_pain() works on Windows", {
+  local_mocked_bindings(Sys.info = function(...) c(sysname = "Windows"))
+  expect_true(is_encoding_a_pain())
+})
+```
+
+The magic happens in `local_mocked_bindings()` calls such as `local_mocked_bindings(Sys.info = function(...) c(sysname = "Windows"))`, that indicates what we want the mocked function, `Sys.info()`, to return in this environment.
```
Comment (hfrick) [inline content/post/2024-03-12-mocking/index.Rmd:116]: This sentence feels too much like a spaghetti sentence to me. Could you break it up?

File: content/post/2024-03-12-mocking/index.Rmd [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+Now, there's a catch here because `Sys.info()` is a base R function.
+We had to add this line in the `R/encoding.R` script:
+
+``` r
+Sys.info <- NULL
+```
+
+This case is documented, as other cases (namespaced calls in particular), in [`testthat::local_mocked_bindings()` manual page](https://testthat.r-lib.org/reference/local_mocked_bindings.html#use).
+
+TODO: example from keyring?
```
Comment (hfrick) [inline content/post/2024-03-12-mocking/index.Rmd:128]: no strong feelings on this one but generally leaning towards "more examples"

File: content/post/2024-03-12-mocking/index.Rmd [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+})
+
+test_that("is_encoding_a_pain() works on Windows", {
+  local_mocked_bindings(Sys.info = function(...) c(sysname = "Windows"))
+  expect_true(is_encoding_a_pain())
+})
+```
+
+The magic happens in `local_mocked_bindings()` calls such as `local_mocked_bindings(Sys.info = function(...) c(sysname = "Windows"))`, that indicates what we want the mocked function, `Sys.info()`, to return in this environment.
+Using testthat's own mocking feels good for three reasons: not needing to add another dependency (and to remember what the mocking packages are called :sweat_smile:), knowing this mocking implementation is [considered better](https://www.tidyverse.org/blog/2023/10/testthat-3-2-0/#mocking) by testthat authors, and getting a [withr vibe](https://withr.r-lib.org/) from the "local\_" aspect.
```
Comment (hfrick) [inline content/post/2024-03-12-mocking/index.Rmd:117]: Since the main motivation for this post was that mocking returns to testthat, this could be a bit more highlighted in the text, e.g., via a bulleted list

```suggestion
Using testthat's own mocking feels good for three reasons: 

- not needing to add another dependency (and to remember what the mocking packages are called :sweat_smile:)
- knowing this mocking implementation is [considered better](https://www.tidyverse.org/blog/2023/10/testthat-3-2-0/#mocking) by testthat authors, and 
- getting a [withr vibe](https://withr.r-lib.org/) from the "local\_" aspect.
```

File: content/post/2024-03-12-mocking/index.Rmd [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+Even the example in a more recent post, on [escape hatches](/2023/01/23/code-switch-escape-hatch-test/), can be easily re-written to use mocking: we'd use `local_mocked_bindings(is_internet_down = function(...) TRUE)`.
+
+Mocking, and its nowadays improved support in testthat, does not completely throw the escape hatch pattern out of the window, but it's still good to keep mocking in mind.
+That post actually also mentioned mocking so it's all good.
+:innocent:
+
+## Conclusion
+
+In this post, we corrected the [someone who was wrong on the internet](https://xkcd.com/386/) years ago.
+In summary, nowadays you can use mocking tools from testthat itself, when you decide mocking is more a help than a hindrance in your testing suite, and when you do not need to turn to specific packages such as [httptest2](https://enpiar.com/httptest2/), [vcr](https://docs.ropensci.org/vcr/) (for HTTP calls), [dittodb](https://docs.ropensci.org/dittodb/) (for database calls).
```
Comment (hfrick) [inline content/post/2024-03-12-mocking/index.Rmd:141]: These special cases are part of the 2019 post but not this one, so they are a little bit of a surprise in a _summary_  for people who only read this post.

**Review by hfrick — COMMENTED** [review]
Nice! Here are my thoughts while reading through it. I read parts of the 2019 post as well but not all the details.

File: content/post/2024-03-12-mocking/index.Rmd (hunk 2 above)
Comment (maelle) [reply content/post/2024-03-12-mocking/index.Rmd]: Thanks, I added a quote from R Packages :smile_cat:

File: content/post/2024-03-12-mocking/index.Rmd (hunk 3 above)
Comment (maelle) [reply content/post/2024-03-12-mocking/index.Rmd]: nope, fixed!
